File: tools/export_template_keys.php
<?php
declare(strict_types=1);

$options = getopt('', ['api-url:', 'template::', 'templateid::', 'out::']);
$api_url = $options['api-url'] ?? '';
$templates = [];
foreach ((array) ($options['template'] ?? []) as $template) {
    $templates[] = $template;
}
foreach ((array) ($options['templateid'] ?? []) as $templateid) {
    $templates[] = '#'.$templateid;
}
$token = getenv('ZBX_TOKEN') ?: '';

if ($api_url === '' || $token === '' || !$templates) {
    fwrite(STDERR, "Usage: ZBX_TOKEN=... php export_template_keys.php --api-url URL --template NAME [--templateid ID]\n");
    exit(2);
}

$rpc_id = 1;
$call = static function (string $method, array $params = []) use ($api_url, $token, &$rpc_id): array {
    $payload = json_encode([
        'jsonrpc' => '2.0',
        'method' => $method,
        'params' => $params,
        'id' => $rpc_id++
    ], JSON_UNESCAPED_SLASHES);

    $curl = curl_init($api_url);
    curl_setopt_array($curl, [
        CURLOPT_POST => true,
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_HTTPHEADER => [
            'Content-Type: application/json-rpc',
            'Authorization: Bearer '.$token
        ],
        CURLOPT_POSTFIELDS => $payload,
        CURLOPT_TIMEOUT => 60,
        CURLOPT_SSL_VERIFYPEER => false,
        CURLOPT_SSL_VERIFYHOST => false
    ]);
    $body = curl_exec($curl);
    if ($body === false) {
        throw new RuntimeException(curl_error($curl));
    }
    $decoded = json_decode($body, true);
    if (!is_array($decoded)) {
        throw new RuntimeException('Invalid JSON response: '.$body);
    }
    if (isset($decoded['error'])) {
        throw new RuntimeException($method.': '.json_encode($decoded['error'], JSON_UNESCAPED_UNICODE));
    }
    return $decoded['result'] ?? [];
};

$resolved = [];
foreach ($templates as $template) {
    if (str_starts_with($template, '#')) {
        $result = $call('template.get', [
            'output' => ['templateid', 'host', 'name'],
            'templateids' => [substr($template, 1)],
            'limit' => 1
        ]);
    }
    else {
        $result = $call('template.get', [
            'output' => ['templateid', 'host', 'name'],
            'filter' => ['host' => [$template]],
            'limit' => 1
        ]);
        if (!$result) {
            $result = $call('template.get', [
                'output' => ['templateid', 'host', 'name'],
                'filter' => ['name' => [$template]],
                'limit' => 1
            ]);
        }
        if (!$result) {
            $result = $call('template.get', [
                'output' => ['templateid', 'host', 'name'],
                'search' => ['host' => $template, 'name' => $template],
                'searchByAny' => true,
                'limit' => 10
            ]);
        }
    }

    foreach ($result as $row) {
        $resolved[$row['templateid']] = $row;
    }
}

$items_by_template = [];
foreach ($resolved as $templateid => $template) {
    $items = $call('item.get', [
        'output' => ['itemid', 'hostid', 'name', 'key_', 'type', 'value_type', 'status'],
        'templateids' => [$templateid],
        'inherited' => true,
        'sortfield' => 'key_'
    ]);

    $items_by_template[$templateid] = [
        'template' => $template,
        'item_count' => count($items),
        'keys' => array_values(array_map(static fn(array $item): array => [
            'name' => $item['name'],
            'key' => $item['key_'],
            'value_type' => $item['value_type'],
            'status' => $item['status']
        ], $items))
    ];
}

$output = [
    'generated_at' => gmdate('c'),
    'templates' => array_values($items_by_template)
];

$json = json_encode($output, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES);
if (($options['out'] ?? '') !== '') {
    file_put_contents($options['out'], $json);
}
echo $json.PHP_EOL;

File: tools/export_template_keys.py
#!/usr/bin/env python3
import argparse
import json
import os
import ssl
import urllib.request


def call(api_url, token, method, params, rpc_id):
    data = json.dumps({
        "jsonrpc": "2.0",
        "method": method,
        "params": params,
        "id": rpc_id,
    }).encode()
    request = urllib.request.Request(
        api_url,
        data=data,
        headers={
            "Content-Type": "application/json-rpc",
            "Authorization": f"Bearer {token}",
        },
        method="POST",
    )
    context = ssl._create_unverified_context()
    with urllib.request.urlopen(request, timeout=60, context=context) as response:
        decoded = json.loads(response.read().decode())
    if "error" in decoded:
        raise RuntimeError(f"{method}: {decoded['error']}")
    return decoded.get("result", [])


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--api-url", required=True)
    parser.add_argument("--template", action="append", default=[])
    parser.add_argument("--templateid", action="append", default=[])
    parser.add_argument("--out")
    args = parser.parse_args()

    token = os.environ.get("ZBX_TOKEN", "")
    if not token:
        raise SystemExit("ZBX_TOKEN is required")

    rpc_id = 1

    def rpc(method, params=None):
        nonlocal rpc_id
        result = call(args.api_url, token, method, params or {}, rpc_id)
        rpc_id += 1
        return result

    resolved = {}
    for templateid in args.templateid:
        rows = rpc("template.get", {
            "output": ["templateid", "host", "name"],
            "templateids": [templateid],
            "limit": 1,
        })
        for row in rows:
            resolved[row["templateid"]] = row

    for name in args.template:
        rows = rpc("template.get", {
            "output": ["templateid", "host", "name"],
            "filter": {"host": [name]},
            "limit": 1,
        })
        if not rows:
            rows = rpc("template.get", {
                "output": ["templateid", "host", "name"],
                "filter": {"name": [name]},
                "limit": 1,
            })
        if not rows:
            rows = rpc("template.get", {
                "output": ["templateid", "host", "name"],
                "search": {"host": name, "name": name},
                "searchByAny": True,
                "limit": 10,
            })
        for row in rows:
            resolved[row["templateid"]] = row

    templates = []
    for templateid, template in sorted(resolved.items(), key=lambda pair: pair[1]["name"]):
        items = rpc("item.get", {
            "output": ["itemid", "hostid", "name", "key_", "type", "value_type", "status"],
            "hostids": [templateid],
            "sortfield": "key_",
        })
        templates.append({
            "template": template,
            "item_count": len(items),
            "keys": [{
                "name": item["name"],
                "key": item["key_"],
                "value_type": item["value_type"],
                "status": item["status"],
            } for item in items],
        })

    payload = {
        "generated_at": __import__("datetime").datetime.utcnow().isoformat() + "Z",
        "templates": templates,
    }
    text = json.dumps(payload, ensure_ascii=False, indent=2)
    if args.out:
        with open(args.out, "w", encoding="utf-8") as file:
            file.write(text)
    print(text)


if __name__ == "__main__":
    main()
